sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,652 @@
+     * When this max read time is not null, the {@link Read} will provide a bounded
+     * {@link PCollection}.
+     */
+    public Read withMaxReadTime(Duration maxReadTime) {
+      checkArgument(maxNumRecords() == Long.MAX_VALUE,
+          "maxNumRecord and maxReadTime are exclusive");
+      return builder().setMaxReadTime(maxReadTime).build();
+    }
+
+    @Override
+    public PCollection<byte[]> expand(PBegin input) {
+      org.apache.beam.sdk.io.Read.Unbounded<byte[]> unbounded =
+          org.apache.beam.sdk.io.Read.from(new RabbitMQSource(this));
+
+      PTransform<PBegin, PCollection<byte[]>> transform = unbounded;
+
+      if (maxNumRecords() != Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(maxNumRecords());
+      } else if (maxReadTime() != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime());
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {
+      checkState(connectionConfig() != null, "RabbitMqIO.read() requires a connection config "
+          + "to be set via withConnectionConfig(config)");
+    }
+
+  }
+
+  static class RabbitMQSource extends UnboundedSource<byte[], RabbitMQCheckpointMark> {
+
+    final Read spec;
+
+    public RabbitMQSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<byte[]> getDefaultOutputCoder() {
+      return ByteArrayCoder.of();
+    }
+
+    @Override
+    public List<RabbitMQSource> split(int desiredNumSplits,
+                                      PipelineOptions options) {
+      // RabbitMQ uses queue, so, we can have several concurrent consumers as source
+      List<RabbitMQSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        sources.add(new RabbitMQSource(spec));
+      }
+      return sources;
+    }
+
+    @Override
+    public UnboundedReader<byte[]> createReader(PipelineOptions options,
+                                                RabbitMQCheckpointMark checkpointMark) {
+      return new UnboundedRabbitMqReader(this, checkpointMark);
+    }
+
+    @Override
+    public Coder<RabbitMQCheckpointMark> getCheckpointMarkCoder() {
+      return SerializableCoder.of(RabbitMQCheckpointMark.class);
+    }
+  }
+
+  static class RabbitMQCheckpointMark
+      implements UnboundedSource.CheckpointMark, Serializable {
+
+    transient Channel channel;
+    Instant oldestTimestamp;
+    final List<String> correlationIds = new ArrayList<>();
+    final List<Long> sessionIds = new ArrayList<>();
+
+    @Override
+    public void finalizeCheckpoint() throws IOException {
+      for (Long sessionId : sessionIds) {
+        channel.basicAck(sessionId, false);
+      }
+      channel.txCommit();
+      oldestTimestamp = Instant.now();
+      correlationIds.clear();
```

Comment:
Clearing this on every checkpoint means that we'll only deduplicate if a message with the same id arrives within a single bundle, which seems very limited. Instead better to override getCurrentRecordId() and requiresDeduping(), and let the runner do the deduping.

Reply:
+1, good idea, thanks.